### DISCUSSION TITLE
Android life cycle question / duplicate process

### DISCUSSION THREAD

**Discussion by hkparker** [fyne-io/fyne#4981] [Q&A] [answered]
**tl;dr: is there any situation in which a Fyne application built for android will have its main process running multiple times on the same device?**

Reason for the question:

In order to prevent multiple instances of my application running on a machine, I use a `.pid` file.  When my application starts, it checks for the existence of the `.pid` file.  If it does not exist, it creates one with the current PID.  If it exists, it checks if the PID that is in that file is running.  If there is no process with that PID, it replaces the file with the current PID.  When the application quits it deletes the `.pid` file.  This setup allows the `.pid` file to not get in the way when an application crash fails to delete the file.

On desktop this has worked fine, and on android it's seemed to work fine as well, except one instance in which my app refused to launch because of the presence of a `.pid` file, and an indication that the process was indeed still running.  I'm checking if the PID in the file is running via a `0` signal to the process, and on android I have no idea if that's going to return without error even if the process is not running, or if another process may be using the PID at that point, etc.  So I don't have a lot of confidence in this technique on this platform.  In `adb shell` while using `run-as` to try to assume the same permissions as my app, I could see the PID of my app with `ps -A`, but got an error when attempting a `kill -0` on it to check for liveness, so I'm surprised it ever returned without error.

If I can just trust that android's lifecycle management stuff is hooked into Fyne such that there's only ever one instance of the main process that's running, and that any goroutines launched from that process are terminated with the app's main process, then I can skip this check on android.  But I wanted to see if there was more depth to this issue before just skipping the check.

**Comment by andydotxyz** [comment, marked as answer]
Android and iOS won't allow multiple instances of an application.

**Reply by hkparker** [reply]
Right, ok so if Fyne isn't doing anything that would break that (not sure it even could), then I can just trust the OS and omit this check on mobile.  What must have happened then is I had some sort of crash (or the app was aggressively killed by the OS), and the technique I'm using to check if a process is running doesn't work on android.
